compat/mswindows.cc

```diff
@@ -189,9 +189,10 @@ getgrnam(char *unused) {
 int
 _free_osfhnd(int filehandle)
 {
-    if (((unsigned) filehandle < SQUID_MAXFD) &&
-            (_osfile(filehandle) & FOPEN) &&
-            (_osfhnd(filehandle) != (long) INVALID_HANDLE_VALUE)) {
+    if (((unsigned)filehandle < SQUID_MAXFD) &&
```

Comment:
Nit: please take advantage of the change to remove the initial c-style cast here.

Reply:
We don't really have a consistent policy when it comes to out-of-scope trivial changes. I propose that we leave that to the discretion of the PR author. It might be inconsistent, but it allows everyone to focus on the things that matter the most

Reply:
> We don't really have a consistent policy when it comes to out-of-scope trivial changes. 

I agree, but I am not sure there is consensus about that.

> I propose that we leave that to the discretion of the PR author.

That is essentially my current approach, with the following caveats:

* Any modified code may still be reviewed/discussed. Reviewers have their discretion as well! In most cases, the reviewer should either fix trivial problems inherited by the modified code from its official version (see the next bullet) or leave them be without creating a change request, but it is impossible to hard-code the best decision logic for every possible situation.
* Once the author decides to touch a line of code, the reviewer may (but does not have to) polish that line of code further if such polishing makes sense (because the line is already a part of the diff). Most of such changes should be limited to formatting and other "mechanical" issues, but it is impossible to hard-code what makes sense for every possible situation.

For example, in this particular case, if Amos did not ask you to start modifying these out-of-scope lines (in ways going beyond trivial formatting changes), then I probably would not suggest undoing these changes (to remove the temptation to modify these lines), but I could have reverted these formatting changes myself (to minimize the diff and to clarify what this PR is actually fixing).

Reply:
> > We don't really have a consistent policy when it comes to out-of-scope trivial changes.
> 
> I agree, but I am not sure there is consensus about that.
> 

We do have a set of style changes that have been agreed to be done in all PRs on touched lines. Updating cast style is one of those.

*If* the lines containing C-style casts remain touched by this PR they should be updated to C++ casts. There should not be any problem/bugs **added* by a correct cast upgrade.


> > I propose that we leave that to the discretion of the PR author.
> 

Notice that is precisely what my OP request did. This style detail was marked as a "nit" (to be fixed at authors discretion) and PR approved regardless.

> That is essentially my current approach, with the following caveats:
> 
>     * Any modified code may still be reviewed/discussed. Reviewers have their discretion as well! In most cases, the reviewer should either fix trivial problems inherited by the modified code from its official version (see the next bullet) or leave them be without creating a change request, but it is impossible to hard-code the best decision logic for every possible situation.
> 

Why do you place this on the reviewers shoulders to perform the fix?
The whole point of style change requests is to make submissions easy for reviewers to a) read and b) identify subtle bugs hiding in the C/C++ syntax layout. Accepting arbitrarily styled code just because the author likes their own coding style better is a bad idea.

>     * Once the author decides to touch a line of code, the reviewer may (but does not have to) polish that line of code further if such polishing makes sense (because the line is already a part of the diff).

Again with placing work on reviewer shoulders. We already have agreement that some changes (eg these cast upgrades) be done in each PR touched line.

> Most of such changes should be limited to formatting and other "mechanical" issues, but it is impossible to hard-code what makes sense for every possible situation.
> 

That "but" being the reason things like this are pointed out as review items, not fixed automagically by the maintenance scripts. It is not a reason to avoid mentioning them in reviews.

> 
> For example, in this particular case, if Amos did not ask you to start modifying these out-of-scope lines (in ways going beyond trivial formatting changes), then 

... then I would be inconsistently applying the agreed policy of touched lines being updated to meet current coding/style guidelines.

Reply:
> > > > Amos: please take advantage of the change to remove the initial c-style cast here.

> > > > Alex: I disagree that this line should be changed (at all).

> > > Francesco: We don't really have a consistent policy when it comes to out-of-scope trivial changes.

> > Alex: I agree, but I am not sure there is consensus about that.
 
> Amos: We do have a set of style changes that have been agreed to be done in all PRs on touched lines. 

I was discussing which lines should (not) be touched in the first place.  I was not discussing style changes on touched lines, but, apparently, we do not have a clear agreement on that either! This exchange appears to reconfirm Francesco's and my opinions regarding "consistent policy".

> Updating cast style is one of those.

Is it? Is there a list of those somewhere? Neither [MergeProcedure](https://wiki.squid-cache.org/MergeProcedure) nor [SquidCodingGuidelines](https://wiki.squid-cache.org/SquidCodingGuidelines) mention the word "cast" AFAICT.

As this very PR illustrates, in many cases, correctly changing casting is far from trivial, may have runtime consequences, and, hence, should not be treated the same way as "mechanical" whitespace formatting, HERE removal, or NULL changes.


> _If_ the lines containing C-style casts remain touched by this PR they should be updated to C++ casts. There should not be any problem/bugs *_added_ by a correct cast upgrade.

@kinkie, that "add no problems" condition is often difficult to satisfy -- there is no single C++ cast that can correctly replace a C-style cast in all cases; one has to pick carefully. Moreover, upgrading a wrong C cast to a wrong C++ cast is a step backwards (i.e. it adds a problem). Thus, I continue to recommend closing this Pandora box by undoing these whitespace changes instead of polishing those lines further.

> > > I propose that we leave that to the discretion of the PR author.
 
> Notice that is precisely what my OP request did. This style detail was marked as a "nit" (to be fixed at authors discretion) and PR approved regardless.

Just to avoid misunderstanding: All of the change requests in this PR are optional. Nobody has insisted on anything yet.

> > That is essentially my current approach, with the following caveats:
> > * Any modified code may still be reviewed/discussed. Reviewers have their discretion as well! In most cases, the reviewer should either fix trivial problems inherited by the modified code from its official version (see the next bullet) or leave them be without creating a change request, but it is impossible to hard-code the best decision logic for every possible situation.
 
> Why do you place this on the reviewers shoulders to perform the fix?

I would not phrase it that way -- the reviewer has an option to do nothing at all. However, if the reviewer wants a trivial fix, then it is often better (lower cost, fewer tensions, fewer out-of-scope arguments, etc.) for the qualified reviewer to apply that trivial fix without a change request rather than filing a change request, waiting for the author to address it (or ask questions about or, or argue with it!), and then validate that the author understood and applied the change request correctly and without triggering more problems. All this overhead is usually completely unnecessary. It is often much better to accomplish the same improvement (i.e. the same code state) by implementing it.

> The whole point of style change requests is to make submissions easy for reviewers to a) read and b) identify subtle bugs hiding in the C/C++ syntax layout.

I doubt that is the whole or even primary point, but we do not need to agree on that. The strategy I recommend does not contradict the statement quoted above.

> Accepting arbitrarily styled code just because the author likes their own coding style better is a bad idea.

This feels like an unrelated point/discussion, but whether that is a bad idea depends on specifics. Until we auto-format or regulate everything, there will always be gray areas where we have to either accept author's choice or argue using such usually-vague concepts as consistency, symmetry, tradition, etc.


> > * Once the author decides to touch a line of code, the reviewer may (but does not have to) polish that line of code further if such polishing makes sense (because the line is already a part of the diff).
 
> Again with placing work on reviewer shoulders. 

Again, I would not phrase it that way -- nobody is forcing the reviewer to do anything here. BTW, writing a high-quality change request is also "work". YMMV, but in my experience, that kind of work is often much harder (and costly) than applying trivial code changes, especially in Squid Project context.
 
> > Most of such changes should be limited to formatting and other "mechanical" issues, but it is impossible to hard-code what makes sense for every possible situation.

> That "but" being the reason things like this are pointed out as review items, not fixed automagically by the maintenance scripts. It is not a reason to avoid mentioning them in reviews.

Agreed -- it is not the reason. There are other/different reasons (as detailed above).
 
> > For example, in this particular case, if Amos did not ask you to start modifying these out-of-scope lines (in ways going beyond trivial formatting changes), then
 
> ... then I would be inconsistently applying the agreed policy of touched lines being updated to meet current coding/style guidelines.

Not if you undid the formatting-only changes instead (for example). However, as we have established, there is no consensus (or sufficient knowledge) about the details of that allegedly "agreed policy", so we should not be wasting time discussing what that policy does or does not imply in this case. We should (re)establish consensus regarding that policy first.

Reply:
@yadij, I believe the current code addresses this change request.

Reply:
Indeed it does.